regression_em/model.py

```diff
@@ -1,18 +1,209 @@
+        :return: Logist arrray.
+        """
+        # Perform clipping to avoid log(0) and zero division.
+        clipped = probs
+        clipped[clipped == 0] = self.epsilon
+        clipped[clipped == 1] = 1 - self.epsilon
+
+        return np.log(clipped / (1 - clipped))
+
+    @staticmethod
+    def calc_responsibility(target_prob: float, ref_prob: float, is_positive: bool) -> float:
+        """Return responsibility to be used in EM algorithm.
+
+        :param target_prob: Probability calculated from the parameters to be updated.
+        :param ref_prob: Reference probability.
+        :param is_positive: If True, the sample has a positive label.
+        :return: Responsibility.
+        """
+
+        if is_positive or target_prob == ref_prob == 1:
+            return 1.0
+        else:
+            return (target_prob * (1 - ref_prob)) / (1 - target_prob * ref_prob)
+
+    def update_responsibilities(self, target_params: RegParam, target_feat: np.ndarray,
+                                ref_params: RegParam, ref_feat: np.ndarray,
+                                labels: Sequence[bool]) -> np.ndarray:
+        """Return responsibilities based on M-step parameters.
+
+        .. Note::
+            target_params and ref_params must be (weight vector, intercept).
+
+        :param target_params: Regression params corresponding to the latent factor to be updated.
+        :param target_feat: Feature matrix corresponding to the latent factor to be updated.
+        :param ref_params: Regression params corresponding to the refered latent factor.
+        :param ref_feat: Feature matrix corresponding to the refered latent factor.
+        :param labels: Sequence of boolean indicating each sample is positivei or negative.
+        :return: List of responsibilities.
+        """
+        # The format of params must be (coef vector, intercept).
+        # targetのパラメーター更新のために更新後のrefでrefの確率を計算
+
+        target_probs = self.calc_probs(target_params[0], target_params[1], target_feat)
+        ref_probs = self.calc_probs(ref_params[0], ref_params[1], ref_feat)
+
+        return np.vectorize(self.calc_responsibility)(target_probs, ref_probs, labels)
+
+    def update_params(self, feat_mat: np.ndarray, responsibilities: np.ndarray, sample_weights) -> Tuple[np.ndarray, float]:
+        """Return fitted Logistic Regression params.
+
+        :param feat_mat: Feature matrix to be used to learn responsibility.
+        :param responsibilities: Sequence of responsibilities calculated at E-step.
+        :return: Updated M-step params.
+        """
+        if self.alpha:
+            reg = Ridge(alpha=self.alpha)
+        else:
+            reg = LinearRegression()
+        reg.fit(feat_mat, self.calc_logits(responsibilities), sample_weights)
+        return reg.coef_, reg.intercept_
+
+    def update_expo_params(self, feat_mat: np.ndarray, responsibilities: np.ndarray, sample_weights) -> Tuple[np.ndarray, float]:
```

Comment:
What is the difference between `update_params` and this?

Reply:
There is no difference between them.
I deleted `update_expo_params`.